### PR TITLE
Add/upgrade domain only

### PR DESCRIPTION
**Pull request by yurynix**
This PR aims to allow users to upgrade their "domain only site" to a regular site with plan.
This PR also fixes a few bugs related to previous "featuregate" work ( the first 2 commits ).

#### Testing instructions
  
Run `git checkout add/upgrade-domain-only` and start your server, or open a [live branch](https://calypso.live/?branch=add/upgrade-domain-only)

Flow:
0. Click "Create a new site" on a "domain only site" ( you might want to create one of those first ).
screen shot 2017-02-07 at 14 30 17

1. Choose a theme
screen shot 2017-02-07 at 14 28 01

2. Choose a plan ( no free plan )
screen shot 2017-02-07 at 14 25 26

3. Checkout ( with discounted price due to existing domain )
screen shot 2017-02-07 at 14 01 11

4. Thank you screen AND Site has a plan + no `is_domain_only` option, that requires the server side patch ( D4297 ).
screen shot 2017-02-07 at 16 48 03

#### Reviews
  
- [x] Code
- [x] Product

### CODE REVIEW

File: client/my-sites/controller.js [hunk 1]
```diff
@@ -142,26 +141,26 @@ function isPathAllowedForDomainOnlySite( pathname, domainName ) {
 
 function onSelectedSiteAvailable( context ) {
 	const selectedSite = sites.getSelectedSite();
-	const state = context.store.getState();
+	const getState = () => context.store.getState();
```
Comment (drewblaisdell) [inline client/my-sites/controller.js]: Could this be `const { getState } = context.store;`?

File: client/signup/steps/plans/index.jsx [hunk 2]
```diff
@@ -46,16 +47,24 @@ class PlansStep extends Component {
...
 			stepSectionName,
 			cartItem,
-			privacyItem,
 			...additionalStepData
-		}, [], { cartItem, privacyItem } );
+		};
+
+		const providedDependencies = { cartItem };
+
+		if ( privacyItem ) {
```
Comment (drewblaisdell) [inline client/signup/steps/plans/index.jsx]: For my own understanding: why is this change needed? Shouldn't a loosely false `privacyItem` be ignored [here](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/step-actions.js#L264)?

File: client/my-sites/controller.js (hunk 1 above)
Comment (yurynix) [inline client/my-sites/controller.js]: I think that your suggestion invalidates the contract we have with the function `getState`,
Our contract states:
1. `getState` call should be parameterless
2. `getState` call should be done on the redux store, i.e. the `this` should be `context.store`

So the other option is:
```js
const getState = context.store.getState.bind( context.store );
```

I don't mind either way, so if you prefer that later that's cool by me.

File: client/signup/steps/plans/index.jsx (hunk 2 above)
Comment (yurynix) [inline client/signup/steps/plans/index.jsx]: I'm not very familiar with the signup framework, so I'm sorry if that was not the proper way to handle that.

The reason that I made it here is because we doing the following:

In [onSelectPlan](https://github.com/Automattic/wp-calypso/pull/11224/files#diff-226cf00033910ff5985fa4e54353707aR67) we call SignupActions.[submitSignupStep](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/actions.js#L28) that dispatches the `SUBMIT_SIGNUP_STEP ` action, that is handled in [SignupDependencyStore](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/dependency-store.js#L58) which performs an [assertion](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/dependency-store.js#L38) that no excessive dependencies are provided by a step. 

 Which results in:
screen shot 2017-02-09 at 10 25 18

Again, I'm not really familiar with the signup, but later we [dispatch](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/dependency-store.js#L60) `SIGNUP_DEPENDENCY_STORE_UPDATE` to redux, that overwrites everything that is 'provided' with the new data, so if we have something extra that is falsy / wrong, we might be [overwriting](https://github.com/Automattic/wp-calypso/blob/master/client/state/signup/dependency-store/reducer.js#L15) something that we shouldn't, so that assertion looks ok to me.

File: client/my-sites/controller.js (hunk 1 above)
Comment (gziolo) [inline client/my-sites/controller.js]: Yes, it looks tricky, because of those `dispatch` calls in the middle of the method.

**Comment by yurynix** [comment]
This currently have the following issue:

1. New user visits the domain-only flow
2. User types a domain and selects it
2.5 In that stage few dependencies, such as `domainItem` are set up for the user at the [Signup dependencies store](https://github.com/Automattic/wp-calypso/blob/master/client/state/signup/dependency-store/reducer.js).
3. User visits the register form step and creates the user.
3.5 Sometime in the process, the [serialization kicks in](https://github.com/Automattic/wp-calypso/blob/024095d874125c0d72a42bdedc4a4856d09695d1/client/state/initial-state.js#L73-L85) and serializes the dependencies
4. When user finishes registering her account, [Signup dependencies store is cleared](https://github.com/Automattic/wp-calypso/blob/master/client/state/signup/dependency-store/reducer.js) - because we done with the flow.
5. Page is [refreshed](https://github.com/Automattic/wp-calypso/blob/fcf789ebb43b7d8fe9a912d8fa82d5bee66573b5/client/signup/wpcom-login-form/index.jsx#L17) to /checkout
6. The deserialization initial state kicks in, and gets the old state before our reset (step 4) was persisted, it still has `domainItem` and other dependencies.
7. After checkout user wants to upgrade his domain using this PR (or any other flow, doesn't really matter)
8. But the user still have stuff from the previous signup flow in the store, which is unexpected by the steps.

File: client/my-sites/controller.js [hunk 3]
```diff
@@ -121,11 +120,11 @@ function renderSelectedSiteIsDomainOnly( reactContext, selectedSite ) {
 			title: i18n.translate( 'This feature is not available for domains' ),
 			line: i18n.translate( 'To use this feature you need to create a site' ),
 			action: i18n.translate( 'Create New Site' ),
-			actionURL: `/start/site-selected/?siteSlug=${ encodeURIComponent( selectedSite.slug ) }`,
+			actionURL: `/start/site-selected/?siteSlug=${ encodeURIComponent( selectedSite.slug ) }&siteId=${ encodeURIComponent( selectedSite.ID ) }`,
```
Comment (stephanethomas) [inline client/my-sites/controller.js]: Just curious, why do we need to encode the site id since this is just a number?

File: client/my-sites/upgrades/domain-management/list/domain-only.jsx [hunk 4]
```diff
@@ -10,13 +10,13 @@ import React, { PropTypes } from 'react';
 import EmptyContent from 'components/empty-content';
 import { domainManagementEdit } from 'my-sites/upgrades/paths';
 
-const DomainOnly = ( { domainName, translate } ) => (
+const DomainOnly = ( { domainName, siteId, translate } ) => (
```
Comment (stephanethomas) [inline client/my-sites/upgrades/domain-management/list/domain-only.jsx]: We should declare this new mandatory property in the prop types.

**Comment by stephanethomas** [comment]
> This currently have the following issue:

This will be fixed by https://github.com/Automattic/wp-calypso/issues/11292.

File: client/my-sites/controller.js (hunk 3 above)
Comment (yurynix) [inline client/my-sites/controller.js]: We should always encode parameters, I guess in that specific case it doesn't matter, but if you don't mind I'd like to keep it.

**Comment by stephanethomas** [comment]
The code looks good and works fine. 👍

**Comment by yurynix** [comment]
This broke the main signup flow, and was reverted.

The problem was when a user chose the free domain (no domain product):
```
Uncaught Error: The dependencies [privacyItem] were listed as provided by the plans step but were not provided by it [ current flow: main ].
```
